Python/bytecodes.c

```diff
@@ -727,8 +728,38 @@ dummy_func(
         #endif
         }
 
+       op(_GUARD_BINARY_OP_EXTEND, (left, right -- left, right)) {
+            PyObject *left_o = PyStackRef_AsPyObjectBorrow(left);
+            PyObject *right_o = PyStackRef_AsPyObjectBorrow(right);
+            assert(INLINE_CACHE_ENTRIES_BINARY_OP == 5);
+            _PyBinaryOpCache *cache = (_PyBinaryOpCache *)(next_instr - INLINE_CACHE_ENTRIES_BINARY_OP);
+            PyBinaryOpSpecializationDescr *descr =
+                (PyBinaryOpSpecializationDescr *)read_void(cache->external_cache);
+            assert(descr && descr->guard);
+            int res = descr->guard(left_o, right_o);
+            EXIT_IF(!res);
+        }
+
+        pure op(_BINARY_OP_EXTEND, (left, right -- res)) {
+            PyObject *left_o = PyStackRef_AsPyObjectBorrow(left);
+            PyObject *right_o = PyStackRef_AsPyObjectBorrow(right);
+            assert(INLINE_CACHE_ENTRIES_BINARY_OP == 5);
+            _PyBinaryOpCache *cache = (_PyBinaryOpCache *)(next_instr - INLINE_CACHE_ENTRIES_BINARY_OP);
+            PyBinaryOpSpecializationDescr *descr =
+                (PyBinaryOpSpecializationDescr *)read_void(cache->external_cache);
+
+            STAT_INC(BINARY_OP, hit);
+
+            PyObject *res_o = descr->action(left_o, right_o);
+            DECREF_INPUTS();
+            res = PyStackRef_FromPyObjectSteal(res_o);
+        }
+
+        macro(BINARY_OP_EXTEND) =
+            _GUARD_BINARY_OP_EXTEND + unused/5 + _BINARY_OP_EXTEND;
```

Comment:
If you want to make `desrc` explicit in both ops, you'll need to rewind the instruction stream.
`unused/1 _GUARD_BINARY_OP_EXTEND + rewind/-4 + _BINARY_OP_EXTEND;`
The negative numbers are not supported, but it is easy enough to add them: https://github.com/python/cpython/pull/117627/files#diff-7a9f9b83d196f419177f6294b4648cc6fdc928bd191edfaf4d96d4fc12030acaR354

Reply:
Can't we just do `op(_GUARD_BINARY_OP_EXTEND, (descr/4, left, right -- descr/4, left, right))`?